### PR TITLE
Add remaining water balance terms to basin.arrow

### PR DESCRIPTION
**Pull request by visr**
Technically these terms can be calculated from flow.arrow and the storage. But since calculating the water balance for a Basin is such a common operation, we include them directly. This also makes it easier to locate water balance errors in time and space, by including the error term.

At some point we may also want to add the instantaneous inflow and outflow to the Basin struct, and update them in `formulate_du!`. This would allow us to reject steps with too large of a balance error, using one of the metrics mentioned in #1271. But for now this only adds them to the output via a callback.

### CODE REVIEW

**Comment by visr** [comment]
@gijsber @DanielTollenaar does this proposal and the change in usage.qmd look ok to you?

File: core/src/write.jl [hunk 1]
```diff
@@ -120,17 +124,27 @@ function basin_table(
     end
 
     time = repeat(data.time[begin:(end - 1)]; inner = nbasin)
+    Δtime_seconds = 0.001 * Dates.value.(diff(data.time))
+    Δtime = repeat(Δtime_seconds; inner = nbasin)
     node_id = repeat(Int32.(data.node_id); outer = ntsteps)
+    storage_rate = Δstorage ./ Δtime
+    error =
+        inflow_rate - outflow_rate - storage_rate + precipitation - evaporation + drainage -
```
Comment (SouthEndMusic) [inline core/src/write.jl:132]: I was a bit confused by the sign of the storage rate, maybe write this as `(all flows except storage_rate) - storage_rate` with the brackets to emphasise which difference you are calculating

File: docs/core/usage.qmd [hunk 2]
```diff
@@ -707,23 +707,31 @@ derivative       | Float64  | -        | -
...
-The initial condition is also written to the file.
+- Results of the storage and level of each Basin, which are instantaneous values;
+- Results of the fluxes on each Basin, which are mean values over the `saveat` intervals.
+  In the time column the start of the period is indicated.
+- The initial condition is written to the file, but the final state is not.
+  It will be placed in a separate output state file in the future.
+- The `inflow_rate` and `outflow_rate` are the sum of the flows into and out of the Basin respectively.
+  The actual flows determine in which term they are counted, not the edge direction.
+- The `storage_rate` is flow that adds to the storage in the Basin, increasing the water level.
+- The `error` is the result of adding all terms of the water balance; `inflow_rate` - `outflow_rate` - `storage_rate` + `precipitation` - `evaporation` + `drainage` - `infiltration`.
```
Comment (SouthEndMusic) [inline docs/core/usage.qmd:718]: Same here

**Review by SouthEndMusic — APPROVED** [review]
Good stuff

**Comment by SouthEndMusic** [comment]
> At some point we may also want to add the instantaneous inflow and outflow to the Basin struct, and update them in `formulate_du!`. This would allow us to reject steps with too large of a balance error, using one of the metrics mentioned in #1271. But for now this only adds them to the output via a callback.

Would it be an idea to integrate the absolute water balance error per basin as a state of the system of differential equations, and throw an error in the same way we do for negative storages if this gets too big? Now that I think of it, why don't we integrate the vertical fluxes this way, just let the solver do it?

**Comment by DanielTollenaar** [comment]
Good stuff! What about considering a `relative_error` as well?

How I compute these in Python (think we have to think about something to avoid replace infinite (balance error devided by 0 in- or outflow_rate):
```
# possible positive error (if error not 0): inflow is larger than outflow
positive = balance_error_df["inflow_rate"] > balance_error_df["outflow_rate"]
 
# if positive error, we devide the error by the outflow_rate (x% too much inflow compared to outflow, or x% too little outflow compared to inflow)
balance_error_df.loc[positive, ["relative_error"]] = (
    balance_error_df[positive]["error"] / balance_error_df[positive]["outflow_rate"]
) * 100
 
# if negative error, we devide the error by the inflow_rate (x% too much outflow compared to inflow, or x% too little inflow compared to outflow)
balance_error_df.loc[~positive, ["relative_error"]] = (
    balance_error_df[~positive]["error"] / balance_error_df[~positive]["inflow_rate"]
) * 100
```

@SouthEndMusic, guess relative_error is something you may use in the solver?

**Comment by visr** [comment]
In 12c77be4fce0838799c72f2be94433fc8db368a0 I renamed `error` to `balance_error` and added `relative_error`, as a fraction, not a percentage.

@DanielTollenaar I added your `relative_error`, though not sure if this is a good general metric. I agree that it is useful for Basins that are in a river with both inflow and outflow. But now if either is 0 then the `relative_error` is 0. Why this particular metric? Perhaps a fraction discrepancy of all inflows (horizontal and vertical) compared to all outflows is more useful. I think this is more similar to the MODFLOW output as posted at the bottom of https://github.com/Deltares/Ribasim/issues/642#issuecomment-1755654305.

@SouthEndMusic I think it is a good idea to let the integrator integrate the flows as well. Do you want to create an issue for it? We should probably do a small test first to get a feeling for it.

**Comment by DanielTollenaar** [comment]
> In [12c77be](https://github.com/Deltares/Ribasim/commit/12c77be4fce0838799c72f2be94433fc8db368a0) I renamed `error` to `balance_error` and added `relative_error`, as a fraction, not a percentage.
> 
> @DanielTollenaar I added your `relative_error`, though not sure if this is a good general metric. I agree that it is useful for Basins that are in a river with both inflow and outflow. But now if either is 0 then the `relative_error` is 0. Why this particular metric? Perhaps a fraction discrepancy of all inflows (horizontal and vertical) compared to all outflows is more useful. I think this is more similar to the MODFLOW output as posted at the bottom of [#642 (comment)](https://github.com/Deltares/Ribasim/issues/642#issuecomment-1755654305).
> 
> @SouthEndMusic I think it is a good idea to let the integrator integrate the flows as well. Do you want to create an issue for it? We should probably do a small test first to get a feeling for it.

@metric all inflows; agreed, that's better.

File: core/src/write.jl [hunk 3]
```diff
@@ -84,31 +84,38 @@ function basin_table(
...
     evaporation = zeros(nrows)
     drainage = zeros(nrows)
     infiltration = zeros(nrows)
+    balance_error = zeros(nrows)
+    relative_error = zeros(nrows)
 
     idx_row = 0
-
-    for vec in vertical_flux.saveval
+    for vec in saved.vertical_flux.saveval
```
Comment (SouthEndMusic) [inline core/src/write.jl:118]: Maybe not use a variable name that is already the name of a function

File: core/src/write.jl [hunk 4]
```diff
@@ -120,17 +127,39 @@ function basin_table(
...
 
     time = repeat(data.time[begin:(end - 1)]; inner = nbasin)
+    Δtime_seconds = 0.001 * Dates.value.(diff(data.time))
+    Δtime = repeat(Δtime_seconds; inner = nbasin)
     node_id = repeat(Int32.(data.node_id); outer = ntsteps)
+    storage_rate = Δstorage ./ Δtime
+
+    for i in 1:nrows
+        storage_flow = storage_rate[i]
+        storage_increase = storage_flow > 0.0 ? storage_flow : 0.0
```
Comment (SouthEndMusic) [inline core/src/write.jl:137]:
```suggestion
        storage_increase = max(storage_flow, 0.0)
```

File: core/src/write.jl [hunk 5]
```diff
@@ -120,17 +127,39 @@ function basin_table(
...
+        storage_decrease = storage_flow > 0.0 ? 0.0 : -storage_flow
```
Comment (SouthEndMusic) [inline core/src/write.jl:138]:
```suggestion
        storage_decrease = min(-storage_flow, 0.0)
```

**Review by SouthEndMusic — APPROVED** [review]
Added some final comments

**Pull request merged**